File: backend/services/auth-service/internal/store/migrate.go
package store

import (
	"context"
	"fmt"
	"io/fs"
	"sort"

	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/tomovu/enchi/services/auth-service/migrations"
)

// RunMigrations áp dụng các file *.up.sql versioned theo thứ tự, ghi nhận vào
// schema_migrations để idempotent. DDL nằm trong file .sql (không tạo bảng ad-hoc trong code).
func RunMigrations(ctx context.Context, pool *pgxpool.Pool) error {
	if _, err := pool.Exec(ctx, `CREATE TABLE IF NOT EXISTS schema_migrations (
		version    TEXT PRIMARY KEY,
		applied_at TIMESTAMPTZ NOT NULL DEFAULT now()
	)`); err != nil {
		return fmt.Errorf("ensure schema_migrations: %w", err)
	}

	files, err := fs.Glob(migrations.FS, "*.up.sql")
	if err != nil {
		return err
	}
	sort.Strings(files)

	for _, name := range files {
		var applied bool
		if err := pool.QueryRow(ctx,
			`SELECT EXISTS(SELECT 1 FROM schema_migrations WHERE version = $1)`, name,
		).Scan(&applied); err != nil {
			return err
		}
		if applied {
			continue
		}
		body, err := migrations.FS.ReadFile(name)
		if err != nil {
			return err
		}
		if _, err := pool.Exec(ctx, string(body)); err != nil {
			return fmt.Errorf("apply migration %s: %w", name, err)
		}
		if _, err := pool.Exec(ctx, `INSERT INTO schema_migrations (version) VALUES ($1)`, name); err != nil {
			return err
		}
	}
	return nil
}

File: backend/services/srs-service/src/db/migrate.ts
import { Pool } from 'pg';

// Migration versioned (tracked trong schema_migrations) — chạy lúc khởi động.
const MIGRATIONS: { version: string; sql: string }[] = [
  {
    version: '0001_init',
    sql: `
      CREATE TABLE IF NOT EXISTS srs_cards (
        user_id       TEXT NOT NULL,
        vocab_id      TEXT NOT NULL,
        ease_factor   REAL NOT NULL DEFAULT 2.5,
        interval_days INTEGER NOT NULL DEFAULT 0,
        repetitions   INTEGER NOT NULL DEFAULT 0,
        due_date      DATE NOT NULL,
        last_reviewed TIMESTAMPTZ,
        PRIMARY KEY (user_id, vocab_id)
      );
      CREATE INDEX IF NOT EXISTS idx_srs_due ON srs_cards (user_id, due_date);
      CREATE TABLE IF NOT EXISTS processed_events (
        event_id     UUID PRIMARY KEY,
        processed_at TIMESTAMPTZ NOT NULL DEFAULT now()
      );
    `,
  },
];

export async function runMigrations(pool: Pool): Promise<void> {
  await pool.query(
    `CREATE TABLE IF NOT EXISTS schema_migrations (
       version TEXT PRIMARY KEY,
       applied_at TIMESTAMPTZ NOT NULL DEFAULT now()
     )`,
  );
  for (const m of MIGRATIONS) {
    const { rowCount } = await pool.query('SELECT 1 FROM schema_migrations WHERE version = $1', [m.version]);
    if (rowCount && rowCount > 0) continue;
    const client = await pool.connect();
    try {
      await client.query('BEGIN');
      await client.query(m.sql);
      await client.query('INSERT INTO schema_migrations (version) VALUES ($1)', [m.version]);
      await client.query('COMMIT');
    } catch (e) {
      await client.query('ROLLBACK');
      throw e;
    } finally {
      client.release();
    }
  }
}
